File: packages/nitrolist/android/src/main/kotlin/com/nitrolist/NitroListStore.kt
package com.nitrolist

import java.util.concurrent.CopyOnWriteArraySet

data class NitroListItem(
  val id: String,
  val templateId: Int,
  val props: Map<String, Any?>,
)

data class NitroListOptions(
  val estimatedItemHeight: Float,
  val overscanScreens: Float,
  val horizontal: Boolean,
  val layout: String,
  val numColumns: Int,
  val columnGap: Float,
  val rowGap: Float,
  val viewabilityConfig: NitroViewabilityConfig,
  val paginationConfig: NitroPaginationConfig,
)

data class NitroViewabilityConfig(
  val windowSize: Int? = null,
  val overscanBefore: Int? = null,
  val overscanAfter: Int? = null,
  val fallbackIndex: Int? = null,
)

data class NitroPaginationConfig(
  val snapEveryItems: Int? = null,
  val snapIndices: List<Int>? = null,
  val initialIndex: Int? = null,
)

data class NitroListState(
  val items: MutableList<NitroListItem>,
  val options: NitroListOptions,
  var viewabilityConfig: NitroViewabilityConfig,
  var paginationConfig: NitroPaginationConfig,
  val measuredHeights: MutableMap<String, Float> = mutableMapOf(),
)

interface NitroListObserver {
  fun onListChanged(handle: Int)
  fun onScrollToIndex(handle: Int, index: Int, animated: Boolean)
}

object NitroListStore {
  private val lists: MutableMap<Int, NitroListState> = mutableMapOf()
  private val visibleRanges: MutableMap<Int, IntRange> = mutableMapOf()
  private val templateRegistry: MutableMap<String, Int> = mutableMapOf()
  private val observers: CopyOnWriteArraySet<NitroListObserver> = CopyOnWriteArraySet()

  fun registerTemplates(map: Map<String, Int>) {
    templateRegistry.putAll(map)
  }

  fun createList(handle: Int, items: List<NitroListItem>, options: NitroListOptions) {
    lists[handle] = NitroListState(
      items = items.toMutableList(),
      options = options,
      viewabilityConfig = options.viewabilityConfig,
      paginationConfig = options.paginationConfig,
    )
    visibleRanges.remove(handle)
    notifyListChanged(handle)
  }

  fun update(handle: Int, patch: List<Map<String, Any?>>) {
    val state = lists[handle] ?: return

    patch.forEach { op ->
      val type = op["op"] as? String ?: return@forEach
      val index = (op["index"] as? Number)?.toInt() ?: return@forEach

      if (type == "remove") {
        if (index in 0 until state.items.size) {
          state.items.removeAt(index)
        }
        return@forEach
      }

      val itemMap = op["item"] as? Map<*, *> ?: return@forEach
      val item = parseItemMap(itemMap) ?: return@forEach

      if (type == "insert") {
        val clamped = index.coerceIn(0, state.items.size)
        state.items.add(clamped, item)
      } else if (type == "update" && index in 0 until state.items.size) {
        state.items[index] = item
      }
    }

    notifyListChanged(handle)
  }

  fun state(handle: Int): NitroListState? = lists[handle]

  fun scrollToIndex(handle: Int, index: Int, animated: Boolean) {
    observers.forEach { it.onScrollToIndex(handle, index, animated) }
  }

  fun configureViewability(handle: Int, config: NitroViewabilityConfig) {
    lists[handle]?.viewabilityConfig = config
  }

  fun configurePagination(handle: Int, config: NitroPaginationConfig) {
    lists[handle]?.paginationConfig = config
  }

  fun dispose(handle: Int) {
    lists.remove(handle)
    visibleRanges.remove(handle)
    notifyListChanged(handle)
  }

  fun updateVisibleRange(handle: Int, first: Int, last: Int) {
    val state = lists[handle] ?: return
    if (state.items.isEmpty()) {
      visibleRanges.remove(handle)
      return
    }

    val maxIndex = state.items.size - 1
    val clampedFirst = first.coerceIn(0, maxIndex)
    val clampedLast = last.coerceIn(clampedFirst, maxIndex)
    visibleRanges[handle] = clampedFirst..clampedLast
  }

  fun viewability(
    handle: Int,
    config: NitroViewabilityConfig? = null,
  ): Map<String, Any?>? {
    val state = lists[handle] ?: return null
    if (state.items.isEmpty()) return null

    val maxIndex = state.items.size - 1
    val resolved = resolveViewabilityConfig(state, config)
    val safeWindow = (resolved.windowSize ?: 1).coerceAtLeast(1)
    val safeBefore = (resolved.overscanBefore ?: 2).coerceAtLeast(0)
    val safeAfter = (resolved.overscanAfter ?: 2).coerceAtLeast(0)

    val visibleRange = visibleRanges[handle] ?: run {
      val first = (resolved.fallbackIndex ?: 0).coerceIn(0, maxIndex)
      val last = (first + safeWindow - 1).coerceIn(first, maxIndex)
      first..last
    }

    val firstVisibleIndex = visibleRange.first.coerceIn(0, maxIndex)
    val lastVisibleIndex = visibleRange.last.coerceIn(firstVisibleIndex, maxIndex)

    val visibleIndices = (firstVisibleIndex..lastVisibleIndex).toList()
    val firstRendered = (firstVisibleIndex - safeBefore).coerceAtLeast(0)
    val lastRendered = (lastVisibleIndex + safeAfter).coerceAtMost(maxIndex)
    val renderedIndices = (firstRendered..lastRendered).toList()
    val visibleSet = visibleIndices.toSet()
    val outsideViewportIndices = renderedIndices.filter { !visibleSet.contains(it) }

    return mapOf(
      "firstVisibleIndex" to firstVisibleIndex,
      "lastVisibleIndex" to lastVisibleIndex,
      "visibleIndices" to visibleIndices,
      "renderedIndices" to renderedIndices,
      "outsideViewportIndices" to outsideViewportIndices,
      "visibleIds" to visibleIndices.map { idx -> state.items[idx].id },
      "renderedIds" to renderedIndices.map { idx -> state.items[idx].id },
      "outsideViewportIds" to outsideViewportIndices.map { idx -> state.items[idx].id },
    )
  }

  fun pagination(handle: Int): Map<String, Any?>? {
    val state = lists[handle] ?: return null
    val itemCount = state.items.size
    val maxIndex = (itemCount - 1).coerceAtLeast(0)
    val currentIndex = visibleRanges[handle]?.first
      ?: state.paginationConfig.initialIndex
      ?: 0
    val clampedIndex = currentIndex.coerceIn(0, maxIndex)
    val snapPoints = resolveSnapPoints(state.paginationConfig, itemCount)
    val snapIndex = nearestSnapIndex(clampedIndex, snapPoints)

    return mapOf(
      "snapIndex" to snapIndex,
      "snapCount" to snapPoints.size,
      "snapPoints" to snapPoints,
      "currentIndex" to clampedIndex,
      "page" to snapIndex,
      "pageCount" to snapPoints.size,
    )
  }

  fun addObserver(observer: NitroListObserver) {
    observers.add(observer)
  }

  fun removeObserver(observer: NitroListObserver) {
    observers.remove(observer)
  }

  fun parseItems(items: List<Map<String, Any?>>): List<NitroListItem> {
    return items.mapNotNull(::parseItemMap)
  }

  private fun parseItemMap(map: Map<*, *>): NitroListItem? {
    val id = map["id"] as? String ?: return null
    val templateId = (map["templateId"] as? Number)?.toInt() ?: return null
    val propsAny = map["props"]
    @Suppress("UNCHECKED_CAST")
    val rawProps = (propsAny as? Map<String, Any?>) ?: emptyMap()
    val props = rawProps.toMutableMap()
    (rawProps["text"] as? String)?.let { props["text"] = it.take(320) }
    (rawProps["cta"] as? String)?.let { props["cta"] = it.take(320) }
    return NitroListItem(id = id, templateId = templateId, props = props)
  }

  private fun resolveViewabilityConfig(
    state: NitroListState,
    override: NitroViewabilityConfig?,
  ): NitroViewabilityConfig {
    return NitroViewabilityConfig(
      windowSize = override?.windowSize ?: state.viewabilityConfig.windowSize,
      overscanBefore = override?.overscanBefore ?: state.viewabilityConfig.overscanBefore,
      overscanAfter = override?.overscanAfter ?: state.viewabilityConfig.overscanAfter,
      fallbackIndex = override?.fallbackIndex ?: state.viewabilityConfig.fallbackIndex,
    )
  }

  private fun resolveSnapPoints(config: NitroPaginationConfig, itemCount: Int): List<Int> {
    if (itemCount <= 0) return listOf(0)

    val explicit = config.snapIndices
    if (!explicit.isNullOrEmpty()) {
      val maxIndex = itemCount - 1
      return explicit.map { it.coerceIn(0, maxIndex) }.distinct().sorted()
    }

    val step = (config.snapEveryItems ?: 1).coerceAtLeast(1)
    val points = mutableListOf<Int>()
    var index = 0
    while (index < itemCount) {
      points.add(index)
      index += step
    }
    return points.ifEmpty { listOf(0) }
  }

  private fun nearestSnapIndex(currentIndex: Int, snapPoints: List<Int>): Int {
    var best = 0
    var bestDistance = Int.MAX_VALUE
    snapPoints.forEachIndexed { index, point ->
      val distance = kotlin.math.abs(currentIndex - point)
      if (distance < bestDistance) {
        best = index
        bestDistance = distance
      }
    }
    return best
  }

  private fun notifyListChanged(handle: Int) {
    observers.forEach { it.onListChanged(handle) }
  }
}

File: packages/nitrolist/ios/NitroListStore.swift
import Foundation

struct NitroNativeListItem {
  let id: String
  let templateId: Int
  let props: [String: Any]
}

struct NitroNativeListOptions {
  let estimatedItemHeight: CGFloat
  let overscanScreens: CGFloat
  let horizontal: Bool
  let layout: String
  let numColumns: Int
  let columnGap: CGFloat
  let rowGap: CGFloat
  let viewabilityConfig: NitroNativeViewabilityConfig
  let paginationConfig: NitroNativePaginationConfig
}

struct NitroNativeViewabilityConfig {
  let windowSize: Int?
  let overscanBefore: Int?
  let overscanAfter: Int?
  let fallbackIndex: Int?
}

struct NitroNativePaginationConfig {
  let snapEveryItems: Int?
  let snapIndices: [Int]?
  let initialIndex: Int?
}

struct NitroNativeListState {
  var items: [NitroNativeListItem]
  var options: NitroNativeListOptions
  var viewabilityConfig: NitroNativeViewabilityConfig
  var paginationConfig: NitroNativePaginationConfig
  var measuredHeights: [String: CGFloat]
}

protocol NitroListStoreObserver: AnyObject {
  func nitroListDidChange(handle: Int)
  func nitroListScrollToIndex(handle: Int, index: Int, animated: Bool)
}

final class NitroListStore {
  static let shared = NitroListStore()

  private var templateRegistry: [String: Int] = [:]
  private var lists: [Int: NitroNativeListState] = [:]
  private var visibleRanges: [Int: ClosedRange<Int>] = [:]
  private var observers: NSHashTable<AnyObject> = NSHashTable.weakObjects()

  private init() {}

  func registerTemplates(_ map: [String: Int]) {
    for (key, value) in map {
      templateRegistry[key] = value
    }
  }

  func createList(handle: Int, items: [NitroNativeListItem], options: NitroNativeListOptions) {
    lists[handle] = NitroNativeListState(
      items: items,
      options: options,
      viewabilityConfig: options.viewabilityConfig,
      paginationConfig: options.paginationConfig,
      measuredHeights: [:]
    )
    visibleRanges.removeValue(forKey: handle)
    notifyDidChange(handle: handle)
  }

  func update(handle: Int, patch: [[String: Any]]) {
    guard var state = lists[handle] else {
      return
    }

    for op in patch {
      guard
        let kind = op["op"] as? String,
        let index = op["index"] as? Int
      else {
        continue
      }

      if kind == "remove" {
        if index >= 0 && index < state.items.count {
          state.items.remove(at: index)
        }
        continue
      }

      guard
        let itemMap = op["item"] as? [String: Any],
        let item = parseItem(itemMap)
      else {
        continue
      }

      if kind == "insert" {
        let clampedIndex = max(0, min(index, state.items.count))
        state.items.insert(item, at: clampedIndex)
      } else if kind == "update", index >= 0, index < state.items.count {
        state.items[index] = item
      }
    }

    lists[handle] = state
    notifyDidChange(handle: handle)
  }

  func state(for handle: Int) -> NitroNativeListState? {
    return lists[handle]
  }

  func scrollToIndex(handle: Int, index: Int, animated: Bool) {
    for case let observer as NitroListStoreObserver in observers.allObjects {
      observer.nitroListScrollToIndex(handle: handle, index: index, animated: animated)
    }
  }

  func configureViewability(handle: Int, config: NitroNativeViewabilityConfig) {
    guard var state = lists[handle] else {
      return
    }
    state.viewabilityConfig = config
    lists[handle] = state
  }

  func configurePagination(handle: Int, config: NitroNativePaginationConfig) {
    guard var state = lists[handle] else {
      return
    }
    state.paginationConfig = config
    lists[handle] = state
  }

  func dispose(handle: Int) {
    lists.removeValue(forKey: handle)
    visibleRanges.removeValue(forKey: handle)
    notifyDidChange(handle: handle)
  }

  func updateVisibleRange(handle: Int, first: Int, last: Int) {
    guard let state = lists[handle], !state.items.isEmpty else {
      visibleRanges.removeValue(forKey: handle)
      return
    }

    let clampedFirst = max(0, min(first, state.items.count - 1))
    let clampedLast = max(clampedFirst, min(last, state.items.count - 1))
    visibleRanges[handle] = clampedFirst...clampedLast
  }

  func viewability(
    handle: Int,
    config: NitroNativeViewabilityConfig?
  ) -> [String: Any]? {
    guard let state = lists[handle], !state.items.isEmpty else {
      return nil
    }

    let resolved = resolveViewabilityConfig(state: state, override: config)
    let safeWindow = max(1, resolved.windowSize ?? 1)
    let safeBefore = max(0, resolved.overscanBefore ?? 2)
    let safeAfter = max(0, resolved.overscanAfter ?? 2)
    let maxIndex = state.items.count - 1

    let visibleRange: ClosedRange<Int>
    if let range = visibleRanges[handle] {
      visibleRange = range
    } else {
      let first = max(0, min(resolved.fallbackIndex ?? 0, maxIndex))
      let last = max(first, min(first + safeWindow - 1, maxIndex))
      visibleRange = first...last
    }

    let firstVisibleIndex = max(0, min(visibleRange.lowerBound, maxIndex))
    let lastVisibleIndex = max(firstVisibleIndex, min(visibleRange.upperBound, maxIndex))

    let visibleIndices = Array(firstVisibleIndex...lastVisibleIndex)
    let firstRendered = max(0, firstVisibleIndex - safeBefore)
    let lastRendered = min(maxIndex, lastVisibleIndex + safeAfter)
    let renderedIndices = Array(firstRendered...lastRendered)

    let visibleSet = Set(visibleIndices)
    let outsideViewportIndices = renderedIndices.filter { !visibleSet.contains($0) }

    let visibleIds = visibleIndices.map { state.items[$0].id }
    let renderedIds = renderedIndices.map { state.items[$0].id }
    let outsideViewportIds = outsideViewportIndices.map { state.items[$0].id }

    return [
      "firstVisibleIndex": firstVisibleIndex,
      "lastVisibleIndex": lastVisibleIndex,
      "visibleIndices": visibleIndices,
      "renderedIndices": renderedIndices,
      "outsideViewportIndices": outsideViewportIndices,
      "visibleIds": visibleIds,
      "renderedIds": renderedIds,
      "outsideViewportIds": outsideViewportIds
    ]
  }

  func pagination(handle: Int) -> [String: Any]? {
    guard let state = lists[handle] else {
      return nil
    }

    let itemCount = state.items.count
    let config = state.paginationConfig
    let maxIndex = max(0, itemCount - 1)
    let currentIndex: Int
    if let range = visibleRanges[handle] {
      currentIndex = max(0, min(range.lowerBound, maxIndex))
    } else {
      currentIndex = max(0, min(config.initialIndex ?? 0, maxIndex))
    }

    let snapPoints = resolveSnapPoints(config: config, itemCount: itemCount)
    let snapIndex = nearestSnapIndex(currentIndex: currentIndex, snapPoints: snapPoints)

    return [
      "snapIndex": snapIndex,
      "snapCount": snapPoints.count,
      "snapPoints": snapPoints,
      "currentIndex": currentIndex,
      "page": snapIndex,
      "pageCount": snapPoints.count
    ]
  }

  func addObserver(_ observer: NitroListStoreObserver) {
    observers.add(observer)
  }

  func removeObserver(_ observer: NitroListStoreObserver) {
    observers.remove(observer)
  }

  private func notifyDidChange(handle: Int) {
    for case let observer as NitroListStoreObserver in observers.allObjects {
      observer.nitroListDidChange(handle: handle)
    }
  }

  private func resolveViewabilityConfig(
    state: NitroNativeListState,
    override: NitroNativeViewabilityConfig?
  ) -> NitroNativeViewabilityConfig {
    return NitroNativeViewabilityConfig(
      windowSize: override?.windowSize ?? state.viewabilityConfig.windowSize,
      overscanBefore: override?.overscanBefore ?? state.viewabilityConfig.overscanBefore,
      overscanAfter: override?.overscanAfter ?? state.viewabilityConfig.overscanAfter,
      fallbackIndex: override?.fallbackIndex ?? state.viewabilityConfig.fallbackIndex
    )
  }

  private func resolveSnapPoints(
    config: NitroNativePaginationConfig,
    itemCount: Int
  ) -> [Int] {
    if itemCount <= 0 {
      return [0]
    }

    if let snapIndices = config.snapIndices, !snapIndices.isEmpty {
      let maxIndex = itemCount - 1
      return Array(Set(snapIndices.map { max(0, min($0, maxIndex)) })).sorted()
    }

    let step = max(1, config.snapEveryItems ?? 1)
    var points: [Int] = []
    var index = 0
    while index < itemCount {
      points.append(index)
      index += step
    }
    return points.isEmpty ? [0] : points
  }

  private func nearestSnapIndex(currentIndex: Int, snapPoints: [Int]) -> Int {
    var best = 0
    var bestDistance = Int.max
    for (index, point) in snapPoints.enumerated() {
      let distance = abs(currentIndex - point)
      if distance < bestDistance {
        best = index
        bestDistance = distance
      }
    }
    return best
  }

  func parseItems(_ value: [[String: Any]]) -> [NitroNativeListItem] {
    return value.compactMap(parseItem)
  }

  private func parseItem(_ map: [String: Any]) -> NitroNativeListItem? {
    guard
      let id = map["id"] as? String,
      let templateId = map["templateId"] as? Int
    else {
      return nil
    }

    let rawProps = map["props"] as? [String: Any] ?? [:]
    var props = rawProps
    if let text = rawProps["text"] as? String {
      props["text"] = String(text.prefix(320))
    }
    if let cta = rawProps["cta"] as? String {
      props["cta"] = String(cta.prefix(320))
    }

    return NitroNativeListItem(id: id, templateId: templateId, props: props)
  }
}
